### ISSUE TITLE
Rethink sync-actions approach

### ISSUE THREAD

**Issue by dferber90** [commercetools/nodejs#355]
### Description

The nodejs-sdk is not fitting all use-cases of the MC for the `sync-action` generation. Is it time to rethink the fundamental approach?

#### No knowledge of `released`
The `nodejs` package operates on `ProductData`, so it can only compute update actions which modify `staged` based on `draft`.  But it can never take `released` (aka `current`)  into account.

This means it can't generate a few update actions:
- When the `draft` we're passing in is equal to `released` we could simply create one update action: `revert`. But as we don't know about `released` we can't do this.
- When a deleted variant is restored (it is in `released` and in `draft`, but not in `staged`) we could create one update action only `restoreVariant`). Since we don't know about `released` we can't possibly know that it existed before.

#### No knowledge of main product data
And then there's the issue that `nodejs` simply doesn't have access to some information:
- We only know about `published` in product projections, but not in a the regular  `product` representation. So we can never compute a `publish` action.

The way out of this would be to not accept `ProductData` but to just accept `Product`.

```js
const actions = nodejs.buildActionsByProduct(
  // pass in "Product", not "ProductData"
  { id: 'p1', published: false, masterData: { current, staged: { x: 1 } } },
  { id: 'p2', published: true, masterData: { current, staged: { x: 2} } },
)

console.log(actions)
// -> [{ action: 'publish' }, { setX: 2 }]
```

#### Does not work with GraphQL at all

For the MC we started using GraphQL for some views. When trying to compute update actions for products loaded from GraphQL we can't do that as the shape is different. So we need to map to the API schema manually. Since CT provides a public GraphQL endpoint we should offer a way to map from the public GraphQL schema to one this library understands.

It could also have different APIs depending on the product representation (`buildActionsByProduct`, `buildActionsByProductData`, `buildActionsByGraphQl`).

I only recently thought of all that, so I'm not too sure where the limitations/downsides of this approach are.

### Possible Solution
It might make more sense to do a rewrite then to change the API so drastically.

Another, unrelated problem is that we're introducing unnecessary generality by using `jsondiffpatch`. We know our schemas, so we don't need to be abstract. We can compare the actual values statically. This makes it hard when comparing across variants, or when trying to compute update actions for reordered `images` on a product as far as I know (but this "gettting rid of `jsondiffpatch` is a different issue).

**Comment by dferber90** [comment]
My goal with this is to raise awareness of the issues we faced when trying to use the package for the modified products view. I don't want to cause any immediate action.

I'd be curious about feedback.

**Comment by tdeekens** [comment]
### Drastic rewrite

We have talked about this quite a bit. I'd like to know the rationale behind using `jsondiffpatch` (or any generic differ) when the decision was taken. Cause I assume it was a well founded one.
I agree that, as we have an abstraction (format) over an generic data format we can have a lot of expectations towards the data we operate on. Essentially making it feel like looping left picking the right and diffing on each entity.

This could mean that we might have an API of which we can expose more narrow differs such as `diffProduct`, diffVariants`, `diffImages` etc. each spitting out update actions only regarding their entity.

### Revertion

I think it's something the API was not build around as it just wasn't a use case back then. Sometimes the consumer might often already know that we wants a revertion which might make diffing kind of obsolete. 

I feel like baking this use case and lifting the entity to product from productdata might cause some trouble for the library. I also wonder how it would work with a notion of a draft as it's inherently not part of a product yet.

**Comment by dferber90** [comment]
> We have talked about this quite a bit. I'd like to know the rationale behind using jsondiffpatch (or any generic differ) when the decision was taken.

+1

> more narrow differs

+1

Yes. And then "compose" them to create the high-level differs.

> I think it's something the API was not build around as it just wasn't a use case back then

Yep. I'm considering whether we should just create something new instead of causing a big breaking change in here.

> I also wonder how it would work with a notion of a draft as it's inherently not part of a product yet.

I think I have this one figured out. At the moment you provide a `draft` as `ProductData`. In the future you would provide the `draft` as `Product`. The `draft` we have right now would live in the new product under  `product.masterData.staged`. I don't see an obstacle here. I had talked about this with JF as well. I can explain in person.

**Comment by tdeekens** [comment]
Ah yeah makes sense. I think I follow so you‘d have „product.masterData.{staged, current, draft}? Whereas when diffing we‘d always diff draft against staged, correct?

**Comment by dferber90** [comment]
No, you'd be diffing two entire [`Products`](http://dev.commercetools.com/http-api-projects-products.html#product) instead of [`ProductData`](http://dev.commercetools.com/http-api-projects-products.html#productdata).

You give it the complete product as-is and the complete product as-desired. The changes are computed based on the `product.masterData.staged` of each product. 

```js
const existingProduct = { id: 'p1', published: false, masterData: { current, staged: { x: 1 } } }
const draftedProduct = { id: 'p1', published: true, masterData: { current, staged: { x: 2 } } }

buildActions(existingProduct, draftedProduct)
```

`buildProduct` now compares `existingProduct.masterData.staged` to `draftedProduct.masterData.staged` to figure out the update actions. But it also compares `existingProduct.published` to `draftedProduct.published` to figure out whether it needs to generate a `publish` update-action.

It can now also check `current` to see whether something was restored etc.
`existingProduct.masterData.current` should always equal `draftedProduct.masterData.current`.


Since it might be overkill to always assume a whole product exists, we might provide lower-level differs (as you mentioned) which then operate on `draftedProduct.masterData.current` (this is `ProductData`) only. They would be able to replace the current `nodejs` sync-action generation.

**Comment by dferber90** [comment]
I changed the title now, this issue is only about the `sync-actions`, not about the complete `nodejs` packages.

**Comment by tdeekens** [comment]
Makes sense.

**Comment by emmenko** [comment]
I find it a bit hard to follow your conversation to be honest, I guess I'm missing a bit of context here.

Also consider that this package was initially meant to be used for updating entities in automatic jobs (server-to-server). I can well image that not all use cases can be covered, in this case we should either extend or write a new package for what you guys are proposing (no breaking changes).

> I'd like to know the rationale behind using jsondiffpatch (or any generic differ) when the decision was taken.

Like I said, the initial use cases were quite "simple": given an updated version of the entity, we needed to generate the proper update actions in order to update the entity (as this is the only way to do that in the API). The `jsondiffpatch` library did a pretty good job of abstracting the detection of the changes away, so we went with that instead of doing all the checks manually.

@dferber90 _PS: remember that this a public repo, I removed the private link from the PR description. Thanks._

**Comment by butenkor** [comment]
> But it can never take released (aka current) into account

One could supply to buildActions existingProduct.current as new version of product and existingProduct.staged as old version. Sync would return update actions like addVariant (equivalent to revert) or change variant actions (aka partial revert like attribute values only)

> So we can never compute a publish action

I am curious what is the use case for computing publish actions? On create it can be supplied within ProductDraft and on update just attached as last update action to actions list. I am asking as publishing is not just an attribute or property to synchronise but rather represents a state which is controller by other manual (product review/approval) or automated processes (bound to attributes validation).

**Comment by dferber90** [comment]
> > But it can never take released (aka current) into account
>
> One could supply to buildActions existingProduct.current as new version of product and existingProduct.staged as old version. Sync would return update actions like addVariant (equivalent to revert) or change variant actions (aka partial revert like attribute values only)

That's not our use-case. `addVariant` is not equivalent to `revert` as the variant would have a new `id`. That is relevant for the ease of use within the MC. Our draft can have changes aside from what's in `staged`.

> > So we can never compute a publish action
>
> I am curious what is the use case for computing publish actions? On create it can be supplied within ProductDraft and on update just attached as last update action to actions list. I am asking as publishing is not just an attribute or property to synchronise but rather represents a state which is controller by other manual (product review/approval) or automated processes (bound to attributes validation).

For Modified Products View: We want to offer a way to save changes to a product while publishing it at the same time. In case the changes we make to the staged version result in the product being equal to the current version, the product is automatically considered published. At this point, when we manually attached a publish-action, the publish update action will fail. So the whole operation will return an error, even though the product ultimately is in the desired state.

For Product Details: Further, we often use this package to determine whether we should show a save-toolbar or not (in update views). The product state dropdown which allows to publish a product should be able to go through the regular flow of updating the draft. Then the sync-actions would tell us that there is at least one update action. Then we can use that information to decide whether to render the save-toolbar or not. At the moment we have to treat the publish-state as an edge-case because it's not covered by the sync-actions.


---

The use-cases within the MC are quite different from the use-cases in import/export it seems. I'm not blaming the `sync-actions` at all, they were created to solve a different problem. But since we use it as well, we should either create a new tool which solves our problems, rethink our approach, or extend this tool. The question is which path to take.

**Comment by nkuehn** [comment]
Just FYI another use case I just discussed with partners to keep in mind: The sync-action builders are also used in deploy and devops automation in the shop solutions to automate the project setup in a "configuration as code" pattern.  

E.g. to get the new product type, tax rates and sometimes even discount rule modeling live in a safe way after having tested it on staging.  types are needed, too btw.

**Comment by StarpTech** [comment]
Hi, what's the status here?

**Comment by StarpTech** [comment]
Some feedback on my experience:

- It's not clear when using `sync-actions` if data is lost e.g when I change the `fieldDefinition` type of a ProductType. 

- It's not clear when using `sync-actions` what data is really synced because the implementation doesn't cover all cases e.g there is no support for custom fields in types or product-types.

From my point of view, that package should be reimplemented from scratch in a new package and it would be great to consider #1429

**Comment by katmatt** [comment]
@StarpTech Thanks for your feedback. 

Have you already seen the terraform provider for commercetools (https://github.com/labd/terraform-provider-commercetools) from our partner labd? This covers already a lot of resource types and this makes it really easy to use the configuration as code paradigm.

We currently don't have any plans to reimplement our `sync-actions` module. But I agree with you that we could improve the coverage and documentation. And I will add some of my thoughts to #1429 too 😄

**Comment by dferber90** [comment]
I'm cleaning up my old issues, so feel free to reopen if this is still relevant.

*Hope you're all having a lovely Sunday at Commercetools :)*
